### PR TITLE
Use new endpoints, decouple list + test editing

### PR DESCRIPTION
**Pull request by tomrf1**
A [previous PR](https://github.com/guardian/support-admin-console/pull/346) added new endpoints enabling the client to:
- lock+edit individual tests in dynamodb
- lock the list ordering using the existing S3 lock file

This PR refactors the client to enable multiple users to work on a channel at the same time.

### UX changes:
- users click a 'Reorder test list' button above the test list if they want to lock/edit it.
- the `StickyBottomBar` has been removed, and replaced with a `StickyTopBar` on the TestEditor (on the right-hand-side). Users now use this to lock/save individual tests.
- It's no longer possible to delete tests - users have to archive them. This simplifies things. In future we can consider cleaning up old archived tests from dynamodb.

### Code changes:
It introduces the `TestsForm` higher-order component, which:
- Renders the Sidebar
- Renders a given `TestEditor` (which is custom for each channel)
- Holds the top-level state and takes care of all interactions with the backend

It also introduces the `ValidatedTestEditor` higher-order component, which:
- Renders the `StickyTopBar` above a given `TestEditor`
- Handles validation - preventing the user from saving invalid data

### Screenshots
Screen Shot 2022-06-10 at 11 08 02
Screen Shot 2022-06-10 at 11 08 08
Screen Shot 2022-06-10 at 11 08 14

### CODE REVIEW

File: public/src/components/channelManagement/bannerTests/bannerTestEditor.tsx [hunk 1]
```diff
@@ -34,36 +34,31 @@ const testCopyHasTemplate = (test: BannerTest, template: string): boolean =>
...
-  onDelete: () => void;
-  onArchive: () => void;
-  onTestSelected: (testName: string) => void;
-  testNames: string[];
-  testNicknames: string[];
-  createTest: (newTest: BannerTest) => void;
-}
-
-const BannerTestEditor: React.FC<BannerTestEditorProps> = ({
+const BannerTestEditor: React.FC<TestEditorProps<BannerTest>> = ({
```
Comment (tomrf1) [inline public/src/components/channelManagement/bannerTests/bannerTestEditor.tsx:37]: Most of this file hasn't changed. It now takes the standard `TestEditorProps` and renders the `StickyTopBar`

File: public/src/components/channelManagement/epicTests/epicTestEditor.tsx [hunk 2]
```diff
@@ -41,368 +41,361 @@ const copyHasTemplate = (test: EpicTest, template: string): boolean =>
...
-  onDelete: () => void;
-  onArchive: () => void;
-  onTestSelected: (testName: string) => void;
-  testNames: string[];
-  testNicknames: string[];
-  testNamePrefix?: string;
-  createTest: (newTest: EpicTest) => void;
-}
-
-const EpicTestEditor: React.FC<EpicTestEditorProps> = ({
```
Comment (tomrf1) [inline public/src/components/channelManagement/epicTests/epicTestEditor.tsx:60]: Most of this file hasn't changed. It now takes the standard `TestEditorProps` and renders the `StickyTopBar`.

File: public/src/components/channelManagement/epicTests/epicTestsForm.tsx [hunk 3]
```diff
@@ -1,214 +1,56 @@
...
-  Variant,
 } from '../helpers/shared';
-import TestsForm, { InnerComponentProps } from '../testEditor';
-import TestsFormLayout from '../testsFormLayout';
-import Sidebar from '../sidebar';
-import { FrontendSettingsType } from '../../../utils/requests';
-import { getDefaultTest, getDefaultVariant } from './utils/defaults';
-import { ControlProportionSettings } from '../helpers/controlProportionSettings';
-
-export enum TickerEndType {
```
Comment (tomrf1) [inline public/src/components/channelManagement/epicTests/epicTestsForm.tsx:30]: all of these models have moved to `/models/epic.ts`, in common with the banner/header models

File: public/src/components/channelManagement/headerTests/headerTestEditor.tsx [hunk 4]
```diff
@@ -20,55 +19,47 @@ import TestEditorVariantSummary from '../testEditorVariantSummary';
...
-  onDelete: () => void;
-  onArchive: () => void;
-  onTestSelected: (testName: string) => void;
-  testNames: string[];
-  testNicknames: string[];
-  createTest: (newTest: HeaderTest) => void;
-}
-
-const HeaderTestEditor: React.FC<HeaderTestEditorProps> = ({
+const HeaderTestEditor: React.FC<TestEditorProps<HeaderTest>> = ({
```
Comment (tomrf1) [inline public/src/components/channelManagement/headerTests/headerTestEditor.tsx:24]: Most of this file hasn't changed. It now takes the standard `TestEditorProps` and renders the `StickyTopBar`

File: public/src/components/channelManagement/stickyTopBar/testArchiveButton.tsx [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    letterSpacing: '1px',
+    color: palette.grey[800],
+  },
+}));
+
+interface TestArchiveButtonProps {
+  onTestArchive: () => void;
+}
+
+export const TestArchiveButton: React.FC<TestArchiveButtonProps> = ({
```
Comment (tomrf1) [inline public/src/components/channelManagement/stickyTopBar/testArchiveButton.tsx:31]: this used to live in `testEditorActionButtons.tsx`

File: public/src/components/channelManagement/stickyTopBar/testCopyButton.tsx [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  existingNames: string[];
+  existingNicknames: string[];
+  sourceName: string;
+  sourceNickname?: string;
+  testNamePrefix?: string;
+  onTestCopy: (oldName: string, newName: string, newNickname: string) => void;
+  disabled: boolean;
+}
+
+export const TestCopyButton: React.FC<TestCopyButtonProps> = ({
```
Comment (tomrf1) [inline public/src/components/channelManagement/stickyTopBar/testCopyButton.tsx:28]: this used to live in `testEditorActionButtons.tsx`

File: public/src/models/epic.ts [hunk 7]
```diff
@@ -0,0 +1,82 @@
+import {
```
Comment (tomrf1) [inline public/src/models/epic.ts:1]: all models here were copied from the old `epicTestsForm.tsx`

File: public/src/components/channelManagement/bannerTests/bannerTestEditor.tsx [hunk 8]
```diff
@@ -34,37 +32,14 @@ const testCopyHasTemplate = (test: BannerTest, template: string): boolean =>
...
-  onDelete: () => void;
-  onArchive: () => void;
-  onTestSelected: (testName: string) => void;
-  testNames: string[];
-  testNicknames: string[];
-  createTest: (newTest: BannerTest) => void;
-}
-
-const BannerTestEditor: React.FC<BannerTestEditorProps> = ({
+const BannerTestEditor: React.FC<ValidatedTestEditorProps<BannerTest>> = ({
```
Comment (tomrf1) [inline public/src/components/channelManagement/bannerTests/bannerTestEditor.tsx:35]: Most of this file hasn't changed. It now takes the standard `ValidatedTestEditorProps`

File: public/src/components/channelManagement/headerTests/headerTestEditor.tsx [hunk 9]
```diff
@@ -6,69 +6,40 @@ import { DeviceType, setStatus, UserCohort } from '../helpers/shared';
...
-  onDelete: () => void;
-  onArchive: () => void;
-  onTestSelected: (testName: string) => void;
-  testNames: string[];
-  testNicknames: string[];
-  createTest: (newTest: HeaderTest) => void;
-}
-
-const HeaderTestEditor: React.FC<HeaderTestEditorProps> = ({
+const HeaderTestEditor: React.FC<ValidatedTestEditorProps<HeaderTest>> = ({
```
Comment (tomrf1) [inline public/src/components/channelManagement/headerTests/headerTestEditor.tsx:22]: Most of this file hasn't changed. It now takes the standard `ValidatedTestEditorProps`

File: public/src/components/channelManagement/headerTests/headerTestVariantEditor.tsx [hunk 10]
```diff
@@ -160,9 +160,9 @@ const HeaderTestVariantContentEditor: React.FC<HeaderTestVariantContentEditorPro
             <div>
               <TextField
                 inputRef={register({ validate: templateValidator })}
-                error={errors.heading !== undefined}
+                error={errors.subheading !== undefined}
```
Comment (tomrf1) [inline public/src/components/channelManagement/headerTests/headerTestVariantEditor.tsx:163]: just fixing a bug I noticed

File: public/src/components/channelManagement/epicTests/epicTestsForm.tsx (hunk 3 above)
Comment (KaliedaRik) [reply public/src/components/channelManagement/epicTests/epicTestsForm.tsx:30]: This refactoring is really welcome - less having to search through the codebase for obscure defs == less tech debt.

File: public/src/components/channelManagement/headerTests/headerTestVariantEditor.tsx (hunk 10 above)
Comment (KaliedaRik) [reply public/src/components/channelManagement/headerTests/headerTestVariantEditor.tsx:163]: (thumbsUpPanda!)

**Review by KaliedaRik — APPROVED** [review]
I very definitely approve of the updates made in the ts/tsx files - especially the refactoring to make navigating the repo a bit easier!

**Review by MaelGNM — APPROVED** [review]
Looks good to me!

**Pull request merged**